FinalClassCheck: NullPointerException on compact source files (JEP 512)

Parent Issue #19971

I have read check documentation: https://checkstyle.org/checks/design/finalclass.html
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

```bash
vivek@Viveks-MacBook-Air checkstyle % cat > config.xml <<'EOF'
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
    "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
    "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
  <module name="TreeWalker">
    <module name="FinalClass"/>
  </module>
</module>
EOF

vivek@Viveks-MacBook-Air checkstyle % cat > InputFinalClassCompact.java <<'EOF'
Runnable r = new Runnable() {
    public void run() { }
};

void main() { }
EOF

vivek@Viveks-MacBook-Air checkstyle %  RUN_LOCALE="-Duser.language=en -Duser.country=US"
java $RUN_LOCALE -jar target/checkstyle-13.8.0-SNAPSHOT-all.jar -c config.xml InputFinalClassCompact.java
Starting audit...
com.puppycrawl.tools.checkstyle.api.CheckstyleException: Exception was thrown while processing InputFinalClassCompact.java
        at com.puppycrawl.tools.checkstyle.Checker.processFiles(Checker.java:313)
        at com.puppycrawl.tools.checkstyle.Checker.process(Checker.java:227)
        at com.puppycrawl.tools.checkstyle.Main.runCheckstyle(Main.java:429)
        at com.puppycrawl.tools.checkstyle.Main.runCli(Main.java:347)
        at com.puppycrawl.tools.checkstyle.Main.execute(Main.java:205)
        at com.puppycrawl.tools.checkstyle.Main.main(Main.java:129)
Caused by: java.lang.NullPointerException: Cannot invoke "com.puppycrawl.tools.checkstyle.checks.design.FinalClassCheck$TypeDeclarationDescription.getQualifiedName()" because the return value of "java.util.Deque.peek()" is null
        at com.puppycrawl.tools.checkstyle.checks.design.FinalClassCheck.visitToken(FinalClassCheck.java:154)
        at com.puppycrawl.tools.checkstyle.TreeWalker.notifyVisit(TreeWalker.java:384)
        at com.puppycrawl.tools.checkstyle.TreeWalker.processIter(TreeWalker.java:457)
        at com.puppycrawl.tools.checkstyle.TreeWalker.walk(TreeWalker.java:322)
        at com.puppycrawl.tools.checkstyle.TreeWalker.processFiltered(TreeWalker.java:202)
        at com.puppycrawl.tools.checkstyle.api.AbstractFileSetCheck.process(AbstractFileSetCheck.java:101)
        at com.puppycrawl.tools.checkstyle.Checker.processFile(Checker.java:341)
        at com.puppycrawl.tools.checkstyle.Checker.processFiles(Checker.java:300)
        ... 5 more
Checkstyle ends with 1 errors.
```

---

**Describe what you expect in detail.**

FinalClassCheck must not crash on a compact source file. A top-level anonymous class declaration in a compact file (a `new Runnable() { ... }` initialiser on a top-level field) is structurally valid under JEP 512 and the check should either process it correctly.

Root cause: in `visitToken`  when the visited token is `LITERAL_NEW` with an anonymous body (last child is `OBJBLOCK`), the check does:

```java
anonInnerClassToOuterTypeDecl
    .put(ast, typeDeclarations.peek().getQualifiedName());
```

`typeDeclarations` is a `Deque<TypeDeclarationDescription>` that is pushed when visiting a `CLASS_DEF` (and the other type-def tokens). For a top-level anonymous class in a compact source file, there is no surrounding `CLASS_DEF` because the anonymous class is a child of a top-level `VARIABLE_DEF` whose parent is `COMPACT_COMPILATION_UNIT`. The Deque is therefore empty, `peek()` returns `null`, and the subsequent `.getQualifiedName()` throws.

Expected on the compact input: zero violations from `FinalClass`.

**Same code inside an ordinary class:**

```bash
vivek@Viveks-MacBook-Air checkstyle % cat > InputFinalClassOrdinary.java <<'EOF'
public class InputFinalClassOrdinary {
    Runnable r = new Runnable() {
        public void run() { }
    };

    public void main() { }
}
EOF

vivek@Viveks-MacBook-Air checkstyle % java $RUN_LOCALE -jar target/checkstyle-13.8.0-SNAPSHOT-all.jar -c config.xml InputFinalClassOrdinary.java

Starting audit...
Audit done.
```

--- a/config/checker-framework-suppressions/checker-nullness-optional-interning-suppressions.xml
+++ b/config/checker-framework-suppressions/checker-nullness-optional-interning-suppressions.xml
@@ -4146,14 +4146,14 @@
     <fileName>src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java</fileName>
     <specifier>dereference.of.nullable</specifier>
     <message>dereference of possibly-null reference typeDeclarations.peek()</message>
-    <lineContent>.put(ast, typeDeclarations.peek().getQualifiedName());</lineContent>
+    <lineContent>outerTypeDeclarationQualifiedName = typeDeclarations.peek().getQualifiedName();</lineContent>
   </checkerFrameworkError>
 
   <checkerFrameworkError unstable="false">
     <fileName>src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java</fileName>
     <specifier>dereference.of.nullable</specifier>
     <message>dereference of possibly-null reference typeDeclarations.peek()</message>
-    <lineContent>outerTypeDeclarationQualifiedName = typeDeclarations.peek().getQualifiedName();</lineContent>
+    <lineContent>outerTypeName = typeDeclarations.peek().getQualifiedName();</lineContent>
   </checkerFrameworkError>
 
   <checkerFrameworkError unstable="false">

--- a/src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java
@@ -150,8 +150,12 @@ public class FinalClassCheck
             case TokenTypes.LITERAL_NEW -> {
                 if (ast.getFirstChild() != null
                         && ast.getLastChild().getType() == TokenTypes.OBJBLOCK) {
-                    anonInnerClassToOuterTypeDecl
-                        .put(ast, typeDeclarations.peek().getQualifiedName());
+
+                    String outerTypeName = packageName;
+                    if (!typeDeclarations.isEmpty()) {
+                        outerTypeName = typeDeclarations.peek().getQualifiedName();
+                    }
+                    anonInnerClassToOuterTypeDecl.put(ast, outerTypeName);
                 }
             }
 
@@ -192,18 +196,18 @@ public class FinalClassCheck
         if (TokenUtil.isTypeDeclaration(ast.getType())) {
             typeDeclarations.pop();
         }
-        if (TokenUtil.isRootNode(ast.getParent())) {
-            anonInnerClassToOuterTypeDecl.forEach(this::registerAnonymousInnerClassToSuperClass);
-            // First pass: mark all classes that have derived inner classes
-            innerClasses.forEach(this::registerExtendedClass);
-            // Second pass: report violation for all classes that should be declared as final
-            innerClasses.forEach((qualifiedClassName, classDesc) -> {
-                if (shouldBeDeclaredAsFinal(classDesc)) {
-                    final String className = CommonUtil.baseClassName(qualifiedClassName);
-                    log(classDesc.getTypeDeclarationAst(), MSG_KEY, className);
-                }
-            });
-        }
+    }
+
+    @Override
+    public void finishTree(DetailAST rootAST) {
+        anonInnerClassToOuterTypeDecl.forEach(this::registerAnonymousInnerClassToSuperClass);
+        innerClasses.forEach(this::registerExtendedClass);
+        innerClasses.forEach((qualifiedClassName, classDesc) -> {
+            if (shouldBeDeclaredAsFinal(classDesc)) {
+                final String className = CommonUtil.baseClassName(qualifiedClassName);
+                log(classDesc.getTypeDeclarationAst(), MSG_KEY, className);
+            }
+        });
     }
 
     /**

--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheckTest.java
@@ -112,6 +112,27 @@ public class FinalClassCheckTest
     }
 
     @Test
+    public void testCompactSourceFile() throws Exception {
+        final String[] expected = {
+            "19:1: " + getCheckMessage(MSG_KEY, "Bar"),
+        };
+        verifyWithInlineConfigParser(
+                getNonCompilablePath("InputFinalClassCompactSourceFile.java"),
+                expected);
+    }
+
+    @Test
+    public void testCompactSourceFileAnonymousNestedResolution() throws Exception {
+        final String[] expected = {
+            "12:5: " + getCheckMessage(MSG_KEY, "Inner"),
+        };
+        verifyWithInlineConfigParser(
+                getNonCompilablePath(
+                        "InputFinalClassCompactSourceFileAnonymousNestedResolution.java"),
+                expected);
+    }
+
+    @Test
     public void testFinalClassConstructorInRecord() throws Exception {
 
         final String[] expected = {

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/design/finalclass/InputFinalClassCompactSourceFile.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/design/finalclass/InputFinalClassCompactSourceFile.java
@@ -1,0 +1,23 @@
+/*
+FinalClass
+
+
+*/
+
+// non-compiled with javac: Compilable with Java25
+
+Runnable r = new Runnable() {
+    public void run() { }
+};
+
+Foo f = new Foo() { };
+
+class Foo {
+    private Foo() { }
+}
+
+class Bar { // violation 'Class Bar should be declared as final'
+    private Bar() { }
+}
+
+void main() { }

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/design/finalclass/InputFinalClassCompactSourceFileAnonymousNestedResolution.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/design/finalclass/InputFinalClassCompactSourceFileAnonymousNestedResolution.java
@@ -1,0 +1,21 @@
+/*
+FinalClass
+
+
+*/
+
+// non-compiled with javac: Compilable with Java25
+
+Inner obj = new Inner() { };
+
+class Outer {
+    static class Inner { // violation 'Class Inner should be declared as final'
+        private Inner() { }
+    }
+}
+
+class Inner {
+    private Inner() { }
+}
+
+void main() { }

